File: scripts/smoke.ts
/** Live smoke: start, hold/send, approve, wait for hello.py + turn_done. */

import { existsSync, readFileSync } from "node:fs";
import { spawn } from "node:child_process";
import { join } from "node:path";
import { ROOT } from "../src/paths.js";
import { readEvents } from "../src/store.js";

const HELLO = "/workspace/hello-cc/hello.py";
const PROMPT =
  "Create or overwrite hello.py that prints Hello, world! then run it with python3.";

function rpc(argv: string[]): Promise<Record<string, unknown>> {
  return new Promise((resolve) => {
    const p = spawn(
      process.execPath,
      [join(ROOT, "node_modules/tsx/dist/cli.mjs"), join(ROOT, "src/cli.ts"), ...argv],
      { cwd: ROOT, stdio: ["ignore", "pipe", "inherit"] },
    );
    let out = "";
    p.stdout.on("data", (c) => {
      out += c.toString();
    });
    p.on("exit", (code) => {
      let obj: Record<string, unknown>;
      const text = out.trim();
      try {
        obj = text ? JSON.parse(text) : { ok: false, error: "empty" };
      } catch {
        obj = { ok: false, error: "non-json", raw: text.slice(0, 500) };
      }
      obj._code = code ?? 1;
      resolve(obj);
    });
  });
}

function kinds(name: string): Set<string> {
  return new Set(readEvents(name).map((e) => String(e.kind)));
}

async function mainSmoke(): Promise<number> {
  console.log("== up");
  console.log(await rpc(["up"]));
  console.log("== up again (idempotent)");
  console.log(await rpc(["up"]));
  console.log("== start smoke");
  const st = await rpc(["start", "--cwd", "/workspace/hello-cc", "--prompt", PROMPT, "--name", "smoke"]);
  const shown: Record<string, unknown> = {};
  for (const k of Object.keys(st)) if (k !== "trace") shown[k] = st[k];
  console.log(shown);
  if (!st.ok) {
    const log = join(ROOT, "data/daemon.log");
    if (existsSync(log)) {
      let tail = readFileSync(log, "utf8").slice(-3000);
      for (const tok of ["ANTHROPIC_AUTH_TOKEN", "sk-ant-", "Bearer "]) {
        if (tail.includes(tok)) {
          tail = "[log redacted: contains secret-like token]";
          break;
        }
      }
      console.log("daemon.log tail:", tail);
    }
    return 1;
  }
  const id = String(st.id);
  console.log("== hold");
  console.log(await rpc(["hold", id]));
  console.log("== send while held (expect error)");
  const held = await rpc(["send", id, "ignore this"]);
  console.log(held);
  if (held.ok || held.error !== "held") {
    console.log("FAIL: expected held error");
    return 1;
  }
  console.log("== release");
  console.log(await rpc(["release", id]));

  const deadline = Date.now() + 240000;
  const approved = new Set<string>();
  let helloOk = false;
  while (Date.now() < deadline) {
    const status = await rpc(["status"]);
    const sessions = Object.fromEntries(
      ((status.sessions as Record<string, unknown>[]) || []).map((s) => [s.id, s]),
    );
    const s = (sessions[id] || {}) as Record<string, unknown>;
    const pending = (s.pending as { tool_use_id?: string; tool?: string; reason?: string }[]) || [];
    for (const p of pending) {
      const tid = p.tool_use_id;
      if (tid && !approved.has(tid)) {
        console.log("approve", tid, p.tool, p.reason);
        console.log(await rpc(["approve", id, tid]));
        approved.add(tid);
      }
    }
    const ks = kinds(id);
    if (existsSync(HELLO)) {
      const text = readFileSync(HELLO, "utf8");
      if (text.includes("Hello")) helloOk = true;
    }
    if (helloOk && (ks.has("turn_done") || ks.has("task_done"))) {
      console.log("== done kinds", [...ks].sort());
      break;
    }
    if (ks.has("dead") && !helloOk) {
      console.log("session dead early", [...ks].sort());
      console.log("last events", readEvents(id).slice(-8));
      return 1;
    }
    await new Promise((r) => setTimeout(r, 1000));
  }
  if (Date.now() >= deadline) {
    console.log("TIMEOUT kinds", [...kinds(id)].sort());
    console.log("status", await rpc(["status"]));
    console.log("events tail", readEvents(id).slice(-15));
    return 1;
  }
  console.log("== hello.py");
  console.log(readFileSync(HELLO, "utf8"));
  console.log("== status");
  const status = await rpc(["status"]);
  console.log(JSON.stringify(status, null, 2).slice(0, 4000));
  const smoke = ((status.sessions as Record<string, unknown>[]) || []).find((x) => x.id === id)!;
  if (!("last_turn" in smoke) || !("last_task" in smoke)) throw new Error("missing last_turn/last_task");
  console.log("== stop");
  console.log(await rpc(["stop", id]));
  console.log("== down");
  console.log(await rpc(["down"]));
  console.log("SMOKE_OK");
  return 0;
}

process.exit(await mainSmoke());

File: scripts/smoke.py
"""Live smoke: start, hold/send, approve, wait for hello.py + turn_done."""

from __future__ import annotations

import json
import os
import sys
import time
from io import StringIO
from pathlib import Path

from ccop.__main__ import main
from ccop.store import read_events

HELLO = Path("/workspace/hello-cc/hello.py")
PROMPT = (
    "Create or overwrite hello.py that prints Hello, world! "
    "then run it with python3."
)


def rpc(argv: list[str]) -> dict:
    buf = StringIO()
    old = sys.stdout
    sys.stdout = buf
    code = 0
    try:
        main(argv)
    except SystemExit as e:
        code = int(e.code or 0)
    finally:
        sys.stdout = old
    text = buf.getvalue().strip()
    try:
        obj = json.loads(text) if text else {"ok": False, "error": "empty"}
    except json.JSONDecodeError:
        obj = {"ok": False, "error": "non-json", "raw": text[:500]}
    obj["_code"] = code
    return obj


def kinds(name: str) -> set[str]:
    return {e["kind"] for e in read_events(name)}


def main_smoke() -> int:
    print("== up")
    print(rpc(["up"]))
    print("== up again (idempotent)")
    print(rpc(["up"]))

    print("== start smoke")
    st = rpc(
        [
            "start",
            "smoke",
            "--cwd",
            "/workspace/hello-cc",
            "--prompt",
            PROMPT,
        ]
    )
    print({k: st[k] for k in st if k != "trace"})
    if not st.get("ok"):
        log = Path("/workspace/ccop/data/daemon.log")
        if log.exists():
            tail = log.read_text(errors="replace")[-3000:]
            # never dump auth tokens
            for tok in ("ANTHROPIC_AUTH_TOKEN", "sk-ant-", "Bearer "):
                if tok in tail:
                    tail = "[log redacted: contains secret-like token]"
                    break
            print("daemon.log tail:", tail)
        return 1

    print("== hold")
    print(rpc(["hold", "smoke"]))
    print("== send while held (expect error)")
    held = rpc(["send", "smoke", "ignore this"])
    print(held)
    if held.get("ok") or held.get("error") != "held":
        print("FAIL: expected held error")
        return 1
    print("== release")
    print(rpc(["release", "smoke"]))

    deadline = time.time() + 240
    approved = set()
    hello_ok = False
    while time.time() < deadline:
        st = rpc(["status"])
        sessions = {s["name"]: s for s in st.get("sessions") or []}
        s = sessions.get("smoke") or {}
        pending = s.get("pending") or []
        for p in pending:
            tid = p.get("tool_use_id")
            if tid and tid not in approved:
                print("approve", tid, p.get("tool"), p.get("reason"))
                print(rpc(["approve", "smoke", tid]))
                approved.add(tid)
        ks = kinds("smoke")
        if HELLO.exists():
            text = HELLO.read_text()
            if "Hello" in text:
                hello_ok = True
        if hello_ok and ("turn_done" in ks or "task_done" in ks):
            print("== done kinds", sorted(ks))
            break
        if "dead" in ks or "failed" in ks:
            # keep going a bit unless dead without progress
            if "dead" in ks and not hello_ok:
                print("session dead early", sorted(ks))
                evs = read_events("smoke")
                print("last events", evs[-8:])
                return 1
        time.sleep(1.0)
    else:
        print("TIMEOUT kinds", sorted(kinds("smoke")))
        print("status", rpc(["status"]))
        print("events tail", read_events("smoke")[-15:])
        return 1

    print("== hello.py")
    print(HELLO.read_text())
    print("== status")
    status = rpc(["status"])
    print(json.dumps(status, indent=2, default=str)[:4000])
    smoke = [x for x in status.get("sessions") or [] if x.get("name") == "smoke"][0]
    assert "last_turn" in smoke and "last_task" in smoke
    print("== stop")
    print(rpc(["stop", "smoke"]))
    print("== down")
    print(rpc(["down"]))
    print("SMOKE_OK")
    return 0


if __name__ == "__main__":
    raise SystemExit(main_smoke())
